File: 20241022/junhee/퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    int start = 1;
    int end = 100000;
    while(start <= end) {
        int mid = (start + end) / 2;
        long long time = times[0];
        for(int i = 1; i < diffs.size(); i++) {
            if(diffs[i] <= mid) time += times[i];
            else {
                int temp = diffs[i] - mid;
                time += (times[i - 1] + times[i]) * temp + times[i];
            }
        }
        if(time <= limit) {
            end = mid - 1;
        }
        else {
            start = mid + 1;
        }
        
    }
    return start;
}

File: 20241021/haeun/퍼즐게임챌린지.js
function solution(diffs, times, limit) {
    const diffLen = diffs.length;

    const answer = binarySearch();

    function binarySearch() {
        let left = 1;
        let right = 100000;

        while (left <= right) {
            const mid = parseInt((left + right) / 2);
            const time = checkTime(mid);

            if (time <= limit) right = mid - 1;
            else left = mid + 1;
        }

        return left;
    }

    function checkTime(level) {
        let totalTime = times[0];
        for (let i = 1; i < diffLen; i++) {
            let k = 0;
            if (diffs[i] > level) {
                k = diffs[i] - level;
            }
            totalTime += (times[i] + times[i - 1]) * k + times[i];
            if (totalTime > limit) return totalTime;
        }

        return totalTime;
    }

    return answer;
}
